src/librustc/ty/context.rs

```diff
@@ -1510,6 +1508,49 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
     {
         self.mk_substs(iter::once(s).chain(t.into_iter().cloned()).map(Kind::from))
     }
+
+    pub fn lint_node<S: Into<MultiSpan>>(self,
+                                         lint: &'static Lint,
+                                         id: NodeId,
+                                         span: S,
+                                         msg: &str) {
+        self.struct_span_lint_node(lint, id, span.into(), msg).emit()
+    }
+
+    pub fn lint_level_at_node(self, lint: &'static Lint, mut id: NodeId)
+        -> (lint::Level, lint::LintSource)
+    {
+        let sets = self.lint_levels(LOCAL_CRATE);
```

Comment:
This creates an edge to the crate-wide `lint_levels` and will force a recompilation of everything with a lint assigned too it whenever the crate-wide `lint_levels` changes, even with red-green. If e.g. you have a ton of functions that are `allow(everything)` and have an allowed lint, you'll have to recompile them every time you add an item.

Until we get red-green this is hard to improve, but if you made `LintLevelMap` a query that would work. I suppose you should add a `// FIXME: red-green` or something.